File: 2111-minimum-operations-to-make-the-array-k-increasing/2111-minimum-operations-to-make-the-array-k-increasing.cpp
class Solution {
public:
    int LIS(vector<int>& vec){
        int n=vec.size();
        
        vector<int> lis; 
        lis.push_back(vec[0]);
        
        for(int i=1; i<n; ++i){
            if(vec[i] >= lis.back()) lis.push_back(vec[i]);
            else{
                auto lb = upper_bound(lis.begin(), lis.end(), vec[i]);
                int idx = lb-lis.begin();
                swap(lis[idx], vec[i]);
            }
        }
        return n-lis.size(); //not part of LIS
    }
    
    int kIncreasing(vector<int>& arr, int k) {
        int n=arr.size();
        int count = 0;
        for(int i=0; i<k; ++i){
            vector<int> vec;
            for(int j=i; j<n; j=j+k){
                vec.push_back(arr[j]);
            }
            count += LIS(vec);
        }
        return count;
    }
};

File: 2111-minimum-operations-to-make-the-array-k-increasing/2111-minimum-operations-to-make-the-array-k-increasing.java
class Solution {
    public int kIncreasing(int[] arr, int k) {
        int kLenLIS= 0, dp[]= new int[arr.length/k+1];
        for(int i=0; i<k; i++)
            kLenLIS+= lengthOfLIS(arr, i, k, dp);
        return arr.length - kLenLIS;
    }
    
    public int lengthOfLIS(int[] arr, int start, int k, int[] dp) {
        int len= 0;
        for(int i=start; i<arr.length; i+=k) {
            int j= upperBound(dp, len, arr[i]);
            if(j==len) len++;
			dp[j]= arr[i];
        }
        return len;
    }
    
    int upperBound(int[] dp, int len, int num){
        int ans= len, l= 0, r= len-1;
        while(l<=r){
            int mid= l+(r-l)/2;
            if(dp[mid]<=num){
                l= mid+1;
            }else{
                ans= mid;
                r= mid-1;
            }
        }
        return ans;        
    }
}